packages/solid/src/render/component.ts

```diff
@@ -265,6 +265,14 @@ export function mergeProps<T extends unknown[]>(...sources: T): MergeProps<T> {
   return target as any;
 }
 
+export type DefaultProps<T, K extends keyof T> = MergeProps<[Required<Pick<T, K>>, T]>;
+export function defaultProps<T, K extends keyof T>(
+  props: T,
+  defaults: Required<Pick<T, K>>
+): DefaultProps<T, K> {
+  return mergeProps(defaults, props);
+}
```

Comment:
```ts
type DefaultableProps<T, K extends keyof T> = {
  [P in K]-?: undefined extends T[P] ? Exclude<T[P], undefined> : never;
}
```

```diff
- Required<Pick<T, K>>
+ DefaultableProps<T, K>
```

I haven't fully tested this but this should allow only defaults that aren't already necessary in `props`.